homeassistant/components/mqtt/sensor.py

```diff
@@ -277,7 +277,9 @@ def _update_last_reset(msg: ReceiveMessage) -> None:
                 )
 
         @callback
-        @write_state_on_attr_change(self, {"_attr_native_value", "_attr_last_reset"})
+        @write_state_on_attr_change(
+            self, {"_attr_native_value", "_attr_last_reset", "_expired"}
+        )
```

Comment:
When a sensor has `expire_after` set, it will set `self._expired` to False. As long as the sensor is not expired, it will stay `False` and hence will not change. If the sensor expires, the state is always saved and `self._expired` is set to `True`. When a new value is received, the timer will be reset and `self._expired` will be set to `False` again.